twill-yarn/src/main/java/org/apache/twill/internal/yarn/YarnUtils.java

```diff
@@ -165,7 +165,7 @@ public static ApplicationId createApplicationId(long timestamp, int id) {
         tokens = ImmutableList.copyOf(fsTokens);
       }
     } else if (factory instanceof FileContextLocationFactory) {
-      FileContext fc = ((FileContextLocationFactory) locationFactory).getFileContext();
+      FileContext fc = ((FileContextLocationFactory) factory).getFileContext();
```

Comment:
There's no end-to-end unit test cases in the twill-yarn module. And there's no unit tests for this class.
Any integration test on a secure cluster would've triggered this failure though.
I've verified that the original issue is no longer hit after I made this change.